### ISSUE TITLE
Collision Filtering and Modification

### ISSUE THREAD

**Issue by datael** [Jondolf/avian#150]
Future features in the readme lists Collision Filtering as something not yet implemented and I would like to get the ball rolling if possible.

### Topics discussed in the xpbd help channel on Discord

Some topics that were discussed on Discord included:
1. Global physics hooks that can iterate over all collisions
2. Per-entity filter components
    * Filtering collisions with entities by EntityId
    * Filtering collisions by predicate
3. Being able to modify the individual contacts
4. The possibility of being able to use one-shot systems in the future
5. The possibility of being able to use entity relations to allow more efficient filtering

### Global Physics hooks
Adding a schedule after the `SubstepSet::NarrowPhase`, and exposing a way to filter collisions from the `Collisions` resource gives a very open-ended and straightforward way to allow for modifying collisions.
I've got a working example [on this fork here](https://github.com/datael/bevy_xpbd/commits/collision_post_processing).

I'm in two minds about whether it's worth the schedule to have built-in sets (e.g. `First`, `Filter`, `Modify`, `Last`), or whether that complexity should be left up to the user. I can see a potential want to pre-filter collisions with faster filters before any slower ones execute, for example.

### Per-entity filter components
@Jondolf suggested something in the following form:
```rust
ContactFilter::new()
        .with_excluded_entities([...])
        .with_predicate(|mut contact: ContactData| {
            // Modify contact (optional)
            contact.penetration = 0.1;
            // Return the contact or use None to ignore contact
            Some(contact)
        }),
```

It was also suggested that some form of combination rule may be needed to handle when two entities with conflicting filter predicates are present.

### Modifying individual contacts
Being able to modify the individual contacts is desirable. Therefore, being able to return `Some(contact)` from a post-processing step is more useful than only being able to remove or keep collisions.

### One-shot systems
@Jondolf showed the following code on Discord as something that it would be nice to be able to do. There was discussion about how this may not be possible with one-shot systems, but using `run_system_with`* may allow it as long as the overhead was acceptable.

\* `run_system_with` was added to bevy with [this merged but unreleased-at-the-time-of-writing pull request](https://github.com/bevyengine/bevy/pull/9366).

```rust
fn setup(mut commands: Commands, contact_hook_pipeline: ContactHookPipeline) {
    commands.spawn((
        ContactHook::new(
            &contact_hook_pipeline,
            |In(mut contact): In<ContactData>, query: Query<&MyBundle>| {
                // Modify contact, query for entities, do whatever
                // Return new contact data to narrow phase; if None, skip contact
                Some(contact)
            }
        ),
    ));
}
```

### Entity Relations

Something along the lines of `(Body1, CollidesWith(Contacts), Body2)` was suggested by @Jondolf.

Bevy currently has a tracking issue for entity relations here: https://github.com/bevyengine/bevy/issues/3742

### Other notes/questions

Should filtering be treated as a separate issue from modification?

My current personal needs for collision filtering are very simplistic: I just want to implement one-way platforms, which is easily solved by the first bullet point above. I therefore have a very narrow view of what needs exist when it comes to being able to filter/modify collisions, and so would like to get input from other interested parties.

One shortfall of [bevy_rapier](https://github.com/dimforge/bevy_rapier) that I would like to avoid was that it seemed impossible to register more than one set of query params to use with collision hooks because the parameters are registered on the plugin itself as a generic type. Being able to access queries while filtering and modifying collisions seems very important in an ECS.

### Where to go from here

I've started work on global filters and entity-entity collision exclusion [in a fork here](https://github.com/datael/bevy_xpbd/commits/collision_filtering), but am looking for input. I'd be interested in creating a pull request at some point, even if it's only for the global collision hook, as it's a small enough change that gives quite a bit of power from the get-go.

**Comment by Jondolf** [comment]
Thanks for such a thorough issue!

I would say that we should kick things off in the initial PR with "global physics hooks/filters" and excluding specific entities. This is probably the easiest and least controversial part, and it should already be enough for a lot of use cases.

Below are some of my current thoughts.

## Global physics hooks

For now, I would just add a custom schedule like you're currently doing. In my opinion, the separate system sets are unnecessary, since users can just use `.after`, `.before` and `.chain` to order the systems or create their own set. The schedule is basically empty by default, so there is nothing to order your systems relative to other than your own systems.

The main caveat with the approach of using a custom schedule and iterating over `Collisions` is that it adds a lot of iteration. We would iterate over all collisions once in the narrow phase, and however many times users iterate over them in their custom filter systems. With Rapier's approach, the filter is applied directly in the narrow phase, so it only iterates through all collisions once per physics frame (+ once in the solver).

The two main solutions I can currently come up with:

1. Rapier's approach of storing physics hooks in a resource and allowing users to implement the filtering logic. Fast, but not very flexible, ergonomic or ECS-like, and only allows one global filter.
2. Something like callbacks based on one-shot systems (again). Allows access to the ECS world, and could allow multiple (ordered) filters if the filtering systems are stored in e.g. a vector. Probably has extra overhead, needs to be benchmarked.

For now though, I think the approach of using a custom schedule like you're doing is perfectly fine for the initial implementation. Even if we eventually implemented some other approach, the filtering could still be done in the custom schedule.

## Excluding specific entities

For excluding specific entities, I recommend moving the logic into the broad phase, next to the collision layer filtering. It's unnecessary and expensive to compute contacts in the narrow phase between entities that will be ignored later.

In terms of the component name, I'm a bit torn. `CollisionFilter` looks nice, but if it only contains excluded entities, some other name might make more sense. We also already have `CollisionLayers`, which is a type of collision filter but is under a separate component, so there might be some confusion.

I would say that it feels more ECS-like to have separate components for different types of filter/hook components, like `ExcludedCollisionEntities` (for the excluded entities), `CollisionHook` (for future predicate-based filtering) and `ActiveCollisionTypes` (for filtering by `RigidBody` type or whether a collider is a sensor, [like Rapier](https://rapier.rs/docs/user_guides/bevy_plugin/colliders#active-collision-types)). This would also play nicer with change detection, and if we really wanted to group the filters for querying ergonomics, we could instead have `CollisionFilter` be a world query. These names are just examples though, I'm open to other names.

## Entity-level contact hooks

Component-based contact hooks/filters using something like one-shot systems would definitely be nice to have, and they would allow for ergonomic entity-specific logic without having to iterate over all of `Collisions`.

However, this will require some experimentation to discover potential caveats and find the best approach. I'm not sure if using callbacks like this is considered good or bad practise in the context of an ECS, as it hasn't really been possible before. I also suspect that one-shot systems would have extra overhead which might be significant if many entities have custom contact hooks.

**Comment by datael** [comment]
Apologies I've been a while getting back to you. I had some personal stuff come up so I haven't had much time to look at this.

Of the things above, I think adding a global filter is the one we're in agreement with, so I've created a pull request just implementing that. It includes an example in 2d of using the filter to implement a one-way platform.

**Comment by Jondolf** [comment]
Hi! It's been a while, but a quick update:

I would consider #610 to be the correct solution for contact filtering and modification. Global collision hooks exist in pretty much all engines I know, and should be flexible enough to efficiently handle the vast majority of advanced contact scenarios (one-way platforms, conveyor belts, non-uniform friction...).

Per-entity contact filtering and modification on the other hand is something that I think is just very tricky / impossible to do efficiently, *if* we want the hooks to have ECS access. To do it without severe performance or usability implications, the hooks must be executed *immediately*, in a potentially parallel loop. It would require exclusive world access, and just isn't very viable here I think. One-shot systems, observers, etc. wouldn't really work.

We *could* just do per-entity hooks that only provide the entity IDs, maybe some collider components (since those might be available in the system already), and mutable access to the contacts, but even then it's not problem-free. What if *both* entities in a collision have hooks? Users might need to take special care to make sure they don't do duplicate work, or to consider the order in which hooks are called, and it quickly becomes a mess. Before doing any work on this, I think I would at least want some really solid practical use cases where per-entity hooks would be useful/necessary, and a reason you couldn't just use global hooks or collision layers for it.

Also FYI, it's pretty likely that `PostProcessCollisions` will be removed in favor of the collision hooks in #610. Running a whole schedule in between contact logic isn't ideal, and seems to prevent some important optimizations I'm working on. Ultimately, I would put performance ahead of the hyper-flexibility of an entire schedule, especially since the hooks should still be enough for basically all scenarios I can think of. I haven't removed it just yet though.

Anyway, just wanted to give my current thoughts on this over a year later. I think I'll close this issue once #610 is merged. New, more narrowly scoped issues can be made afterwards if there are still open questions.
